File: JAVA/tcp-cli-srv/TcpSrvSum.java
import java.io.*; 
import java.net.*; 

class TcpSrvSum {    
	static ServerSocket sock;

	public static void main(String args[]) throws Exception {       
		Socket cliSock;

		try { sock = new ServerSocket(9999); }
		catch(IOException ex) {
			System.out.println("Failed to open server socket");
			System.exit(1);
			}

		while(true) {
			cliSock=sock.accept();
			new Thread(new TcpSrvSumThread(cliSock)).start();
			}
		} 
	}



class TcpSrvSumThread implements Runnable {    
	private Socket s;
	private DataOutputStream sOut;
	private DataInputStream sIn;

	public TcpSrvSumThread(Socket cli_s) { s=cli_s;}

	public void run() {
		long f,i,num,sum;
		InetAddress clientIP;

		clientIP=s.getInetAddress();
		System.out.println("New client connection from " + clientIP.getHostAddress() + 
			", port number " + s.getPort());
		try {
			sOut = new DataOutputStream(s.getOutputStream());
			sIn = new DataInputStream(s.getInputStream());
			do {
   				sum=0;
				do {
   					num=0; f=1; for(i=0;i<4;i++) {num=num+f*sIn.read(); f=f*256;}
					sum=sum+num;
					}
   				while(num>0);
   				num=sum; for(i=0;i<4;i++) {sOut.write((byte)(num%256)); num=num/256; }
   				}
			while(sum>0);

			System.out.println("Client " + clientIP.getHostAddress() + ", port number: " + s.getPort() + 
				" disconnected");
			s.close();
			}
		catch(IOException ex) { System.out.println("IOException"); }
		}
	}

File: C/tcp-cli-srv/tcp_srv_sum.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define BUF_SIZE 300
#define SERVER_PORT "9999"

int main(void) {
	struct sockaddr_storage from;
	int err, newSock, sock;
	unsigned int adl;
	unsigned long i, f, n, num, sum;
	unsigned char bt;
	char cliIPtext[BUF_SIZE], cliPortText[BUF_SIZE];
	struct addrinfo  req, *list;

	bzero((char *)&req,sizeof(req));
	// requesting a IPv6 local address will allow both IPv4 and IPv6 clients to use it
	req.ai_family = AF_INET6;
	req.ai_socktype = SOCK_STREAM;
	req.ai_flags = AI_PASSIVE;      // local address

	err=getaddrinfo(NULL, SERVER_PORT , &req, &list);

	if(err) {
        	printf("Failed to get local address, error: %s\n",gai_strerror(err)); exit(1); }

	sock=socket(list->ai_family,list->ai_socktype,list->ai_protocol);
	if(sock==-1) {
        	perror("Failed to open socket"); freeaddrinfo(list); exit(1);}

	if(bind(sock,(struct sockaddr *)list->ai_addr, list->ai_addrlen)==-1) {
        	perror("Bind failed");close(sock);freeaddrinfo(list);exit(1);}

	freeaddrinfo(list);

	listen(sock,SOMAXCONN);

	puts("Accepting TCP connections (both over IPv6 or IPv4). Use CTRL+C to terminate the server");

	adl=sizeof(from);
	for(;;) {
        	newSock=accept(sock,(struct sockaddr *)&from,&adl);
        	if(!fork()) {
                	close(sock);
			getnameinfo((struct sockaddr *)&from,adl,cliIPtext,BUF_SIZE,cliPortText,BUF_SIZE,
				NI_NUMERICHOST|NI_NUMERICSERV);
                	printf("New connection from node %s, port number %s\n", cliIPtext, cliPortText);
                	do {
                    		sum=0;
                    		do {
		       			num=0;f=1;
					for(i=0;i<4;i++) {
						read(newSock,&bt,1); num=num+bt*f; f=256*f;
						}
                       			sum=sum+num;
                       			}
                    		while(num);
		    		n=sum;
		    		for(i=0;i<4;i++) {
                        		bt=n%256; write(newSock,&bt,1); n=n/256; }
                    		} 
                	while(sum);
                	close(newSock);
                	printf("Connection from node %s, port number %s closed\n", cliIPtext, cliPortText);
                	exit(0);
                	}
        	close(newSock);
        	}
	close(sock);
	}

